### PR TITLE
SwerveConstants, CalibrationFile -> module

### PR DESCRIPTION
**Pull request by YHerm**

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/swerve/SwerveCommandsBuilder.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+			.withName("Drive with supplier state");
+	}
+
+	public Command driveState(DoubleSupplier xSupplier, DoubleSupplier ySupplier, DoubleSupplier rotationSupplier, SwerveState state) {
+		return new InitExecuteCommand(
+			swerve::resetPIDControllers,
+			() -> swerve.driveByState(xSupplier.getAsDouble(), ySupplier.getAsDouble(), rotationSupplier.getAsDouble(), state),
+			swerve
+		).withName("Drive with state");
+	}
```
Comment (drorta) [inline src/main/java/frc/robot/subsystems/swerve/SwerveCommandsBuilder.java:138]: consider driveByState

File: src/main/java/frc/robot/subsystems/swerve/SwerveType.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+	SWERVE("");
+
+	private final String swerveName;
+
+	SwerveType(String swerveName) {
+		this.swerveName = swerveName;
+	}
+
+	public String getLogPath() {
+		return "Subsystems/" + swerveName + "Swerve/";
```
Comment (drorta) [inline src/main/java/frc/robot/subsystems/swerve/SwerveType.java:14]: consider 
```suggestion
		return "Subsystems/" + "Swerve/" + swerveName;
```

File: src/main/java/frc/robot/subsystems/swerve/SwerveType.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package frc.robot.subsystems.swerve;
+
+public enum SwerveType {
+
+	SWERVE("");
```
Comment (drorta) [inline src/main/java/frc/robot/subsystems/swerve/SwerveType.java:5]: why do you need the ""?

File: src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/AimAssistMath.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+public class AimAssistMath {
+
+	public static ChassisSpeeds getRotationAssistedChassisSpeeds(
+		ChassisSpeeds speeds,
+		Rotation2d robotHeading,
+		Rotation2d targetHeading,
+		SwerveConstants swerveConstants
+	) {
+		Rotation2d pidVelocityPerSecond = Rotation2d
```
Comment (drorta) [inline src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/AimAssistMath.java:21]: maybe pidoutput

File: src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/AimAssistMath.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+	public static ChassisSpeeds getObjectAssistedSpeeds(
+		ChassisSpeeds speeds,
+		Pose2d robotPose,
+		Translation2d objectTranslation,
+		SwerveConstants swerveConstants,
+		SwerveState swerveState
+	) {
+		Translation2d noteRelativeToRobot = SwerveMath.getRelativeTranslation(robotPose, objectTranslation);
+		double pidHorizontalVelocityMetersPerSecond = swerveConstants.yMetersPIDController().calculate(0, noteRelativeToRobot.getY());
```
Comment (drorta) [inline src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/AimAssistMath.java:42]: also output here

File: src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/SwerveStateHelper.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+public class SwerveStateHelper {
+
+	private final Swerve swerve;
+	private final SwerveConstants swerveConstants;
+	private final Supplier<Optional<Pose2d>> robotPoseSupplier;
+	private final Supplier<Optional<Translation2d>> noteTranslationSupplier;
+
+	public SwerveStateHelper(
+		Supplier<Optional<Pose2d>> robotPoseSupplier,
+		Supplier<Optional<Translation2d>> noteTranslationSupplier,
```
Comment (drorta) [inline src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/SwerveStateHelper.java:25]: note? cant it be a generic "object"? also in all other places

File: src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/SwerveStateHelper.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+			return isLeft ? RotateAxis.FRONT_LEFT_MODULE : RotateAxis.FRONT_RIGHT_MODULE;
+		}
+		if (Math.abs(currentAllianceHeading.getDegrees()) >= MathConstants.EIGHTH_CIRCLE.getDegrees() * 3) { // -135 - x - 135
+			return isLeft ? RotateAxis.BACK_RIGHT_MODULE : RotateAxis.BACK_LEFT_MODULE;
+		}
+		if (currentAllianceHeading.getDegrees() > 0) { // 45 <= x <= 135
+			return isLeft ? RotateAxis.FRONT_RIGHT_MODULE : RotateAxis.BACK_RIGHT_MODULE;
+		}
+		return isLeft ? RotateAxis.BACK_LEFT_MODULE : RotateAxis.FRONT_LEFT_MODULE; // -45 >= x >= -135
+	}
```
Comment (drorta) [inline src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/SwerveStateHelper.java:63]: can you explain to me what does it do and why do we need it? just asking

File: src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/SwerveStateHelper.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+			case BACK_RIGHT_MODULE -> swerve.getModules().getModule(ModuleUtils.ModulePosition.BACK_RIGHT).getPositionFromCenterMeters();
+		};
+	}
+
+	public RotateAxis getFarRotateAxis(boolean isLeft) {
+		Rotation2d currentAllianceHeading = swerve.getAllianceRelativeHeading();
+		if (Math.abs(currentAllianceHeading.getDegrees()) <= MathConstants.EIGHTH_CIRCLE.getDegrees()) { // -45 <= x <= 45
+			return isLeft ? RotateAxis.FRONT_LEFT_MODULE : RotateAxis.FRONT_RIGHT_MODULE;
+		}
+		if (Math.abs(currentAllianceHeading.getDegrees()) >= MathConstants.EIGHTH_CIRCLE.getDegrees() * 3) { // -135 - x - 135
```
Comment (drorta) [inline src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/SwerveStateHelper.java:56]: consider editing the comment to <= or >= like the other comments

**Review by drorta — CHANGES_REQUESTED** [review]
Nice job!!!!

File: src/main/java/frc/robot/subsystems/swerve/HeadingStabilizer.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+		headingController.enableContinuousInput(-MathConstants.HALF_CIRCLE.getDegrees(), MathConstants.HALF_CIRCLE.getDegrees());
+	}
+
+	public void lockTarget() {
+		targetLocked = true;
+	}
+
+	public void unlockTarget() {
+		targetLocked = false;
+	}
```
[the remark this answers is not in the document]
Comment (YHerm) [reply src/main/java/frc/robot/subsystems/swerve/HeadingStabilizer.java:27]: Less Cool

File: src/main/java/frc/robot/subsystems/swerve/HeadingStabilizer.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+	public void setTargetHeading(Rotation2d targetHeading) {
+		if (!targetLocked) {
+			headingController.reset();
```
[the remark this answers is not in the document]
Comment (YHerm) [reply src/main/java/frc/robot/subsystems/swerve/HeadingStabilizer.java:31]: Need what?

[the remark this answers is not in the document]
Comment (drorta) [reply src/main/java/frc/robot/subsystems/swerve/HeadingStabilizer.java:31]: .reset

[the remark this answers is not in the document]
Comment (YHerm) [reply src/main/java/frc/robot/subsystems/swerve/HeadingStabilizer.java:31]: to reset the i accumolator

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+		Translation2d[] modulePositionsFromCenterMeters = modules.getModulePositionsFromCenterMeters();
+		double sum = 0;
+		for (Translation2d modulePositionFromCenterMeters : modulePositionsFromCenterMeters) {
+			sum += modulePositionFromCenterMeters.getDistance(new Translation2d());
+		}
+		return sum / modulePositionsFromCenterMeters.length;
+	}
+
+	public Rotation2d getGyroAbsoluteYaw() {
+		double inputtedHeadingRadians = MathUtil.angleModulus(gyroSignals.yawSignal().getLatestValue().getRadians());
```
[the remark this answers is not in the document]
Comment (YHerm) [reply src/main/java/frc/robot/subsystems/swerve/Swerve.java:192]: give the gtyro absolute yaw
with angle modulus

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+	}
+
+	private void logNumberOfOdometrySamples() {
+		Logger.recordOutput(getLogPath() + "OdometrySamples", getNumberOfOdometrySamples());
+	}
+
+
+	public int getNumberOfOdometrySamples() {
+		return Math.min(gyroSignals.yawSignal().asArray().length, modules.getNumberOfOdometrySamples());
+	}
```
[the remark this answers is not in the document]
Comment (YHerm) [reply src/main/java/frc/robot/subsystems/swerve/Swerve.java:165]: because its good to know how many updates i give and get

File: src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/SwerveStateHelper.java (hunk 7 above)
Comment (YHerm) [reply src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/SwerveStateHelper.java:63]: What?: gives the left/right far modules from the driver
Why?: for rotations around the module instead of the debtor of the robot (if we learn how to use it can be good vs defence in games)

File: src/main/java/frc/robot/subsystems/swerve/HeadingStabilizer.java [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+			headingController.setSetpoint(targetHeading.getDegrees());
+		}
+	}
```
[the remark this answers is not in the document]
Comment (YHerm) [reply src/main/java/frc/robot/subsystems/swerve/HeadingStabilizer.java:34]: i dont think so, i set when locked a lot of times, to make it easier to handle. look in swerve on handleHeadingControl

File: .git-blame-ignore-revs
[the remark this answers is not in the document]
Comment (YHerm) [reply .git-blame-ignore-revs:1]: dont know

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+
+	public Rotation2d getGyroAbsoluteYaw() {
+		double inputtedHeadingRadians = MathUtil.angleModulus(gyroSignals.yawSignal().getLatestValue().getRadians());
+		return Rotation2d.fromRadians(inputtedHeadingRadians);
+	}
+
+	public Rotation2d getAbsoluteHeading() {
+		double inputtedHeadingRadians = MathUtil.angleModulus(headingSupplier.get().getRadians());
+		return Rotation2d.fromRadians(inputtedHeadingRadians);
+	}
```
[the remark this answers is not in the document]
Comment (YHerm) [reply src/main/java/frc/robot/subsystems/swerve/Swerve.java:199]: ?

File: .git-blame-ignore-revs
[the remark this answers is not in the document]
Comment (drorta) [reply .git-blame-ignore-revs:1]: Than delete it or something idk

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (drorta) [reply src/main/java/frc/robot/subsystems/swerve/Swerve.java:199]: Same as the gyro absolute value, why do you need angle modulus? And what does angle modulus do?

File: src/main/java/frc/robot/subsystems/swerve/SwerveType.java (hunk 3 above)
Comment (YHerm) [reply src/main/java/frc/robot/subsystems/swerve/SwerveType.java:5]: for now we have one swerve so the name should be swerve

File: src/main/java/frc/robot/subsystems/swerve/SwerveType.java (hunk 2 above)
Comment (drorta) [reply src/main/java/frc/robot/subsystems/swerve/SwerveType.java:14]: Oh ok

File: src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/SwerveStateHelper.java (hunk 8 above)
Comment (YHerm) [reply src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/SwerveStateHelper.java:56]: no because its not <= >=

Comment (drorta) [reply src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/SwerveStateHelper.java:56]: So it always returns -135 minus x minus 135?

File: .git-blame-ignore-revs
[the remark this answers is not in the document]
Comment (YHerm) [reply .git-blame-ignore-revs:1]: @emilyyyylime

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (YHerm) [reply src/main/java/frc/robot/subsystems/swerve/Swerve.java:199]: if i have angle between -∞ to ∞ it makes it between -180 and 180

**Review by NoamRosenberg08 — APPROVED** [review]

**Review by Nitay4 — APPROVED** [review]

File: src/main/java/frc/JoysticksBindings.java [hunk 15]
```diff
@@ -1,8 +1,8 @@
 package frc;
 
-import frc.robot.Robot;
```
Comment (Sh-Pe) [inline src/main/java/frc/JoysticksBindings.java:3]: Why?

File: src/main/java/frc/JoysticksBindings.java [hunk 16]
```diff
@@ -1,8 +1,8 @@
 package frc;
```
Comment (Sh-Pe) [inline src/main/java/frc/JoysticksBindings.java:1]: you don't need .git-blame-ignore-revs if it's empty

File: src/main/java/frc/robot/subsystems/swerve/Calibrations.md [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+- PathPlanner RealTimeConstraints
+
+Gyro:
+-----------------------
+- ID
+- Mount pose
+
+Modules:
+-----------------------
+[See module calibrations file:](module/Calibrations.md)
```
Comment (Sh-Pe) [inline src/main/java/frc/robot/subsystems/swerve/Calibrations.md:18]:
```suggestion
[See module calibrations file](module/Calibrations.md)
```

File: src/main/java/frc/robot/subsystems/swerve/SwerveConstants.java [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+public record SwerveConstants(
+	String logPath,
+	String stateLogPath,
+	String velocityLogPath,
+	double velocityAt12VoltsMetersPerSecond,
+	Rotation2d maxRotationalVelocityPerSecond,
+	PIDController xMetersPIDController,
+	PIDController yMetersPIDController,
+	PIDController rotationDegreesPIDController,
+	PPHolonomicDriveController ppHolonomicDriveController
```
Comment (Sh-Pe) [inline src/main/java/frc/robot/subsystems/swerve/SwerveConstants.java:19]:
```suggestion
	PPHolonomicDriveController pathplannerHolonomicDriveController
```

File: src/main/java/frc/robot/subsystems/swerve/SwerveConstants.java [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+import com.pathplanner.lib.path.PathConstraints;
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.math.geometry.Rotation2d;
+import frc.constants.MathConstants;
+
+public record SwerveConstants(
+	String logPath,
+	String stateLogPath,
+	String velocityLogPath,
+	double velocityAt12VoltsMetersPerSecond,
```
Comment (Sh-Pe) [inline src/main/java/frc/robot/subsystems/swerve/SwerveConstants.java:14]: Why exactly 12 volts? I think it'll be a bit more clear to pass it with proper SI and then multiply it.

**Review by Sh-Pe — COMMENTED** [review]
this is just a bunch of relatively random things, but I'm fine wiht it.

File: src/main/java/frc/robot/subsystems/swerve/SwerveConstants.java (hunk 19 above)
Comment (YHerm) [reply src/main/java/frc/robot/subsystems/swerve/SwerveConstants.java:14]: 12 is the default battery voltage permitted in frc

Comment (Sh-Pe) [reply src/main/java/frc/robot/subsystems/swerve/SwerveConstants.java:14]: I think it should be clear from the name. And, those kinds of things could change in the future, hence breaks OCP.
